### ISSUE TITLE
EQ controls?

### ISSUE THREAD

**Issue by Be-ing** [agfline/RME-Fireface-UC-Drivers#1]
Hello @agfline,
I am working on [reverse engineering TotalMix FX for the Babyface Pro](https://linuxmusicians.com/viewtopic.php?p=144074#p144074) to write an application for Linux to use in place of TotalMix FX. With some help from your work and https://github.com/MrBollie/RME-Babyace-Pro-ALSA-Mixer-Patch I have figured out the USB control messages for the Babyface Pro. However, I am stumped about how to control the onboard EQ and low cut DSP. In Wireshark, I see USB bulk messages being sent when manipulating the EQs in TotalMix FX on Windows. However, that bulk endpoint (10) is in a vendor defined class interface which does not exist in the USB descriptors for class compliant mode, so I don't know how to control those in CC mode. I [asked RME about this](https://forum.rme-audio.de/viewtopic.php?id=35220) and they confirmed it is possible to control the EQs in class compliant mode, but they did not say how. Do you have any ideas how to do this?

**Comment by agfline** [comment]
Hi Be-ing,

Making a generic audio interface controller was also one of my goals when I started porting the fireface UC. If you make it with enough abstraction and modularity, you could achieve something that would work with RME products and many more like Focusrites. That would be such a progress for the linux professional audio world !

That being said, the RME fireface UC i've been working on has very limited DSP capacities with no EQ, thus I don't know much about it. However it is possible that the endpoint number changes in CC mode. Can you post your Babyface Pro device descriptors, both in CC mode and vendor-driver mode so I can have a look ?

**Comment by Be-ing** [comment]
I got a [response](https://forum.rme-audio.de/viewtopic.php?pid=186753#p186753) from RME saying that TotalMix FX for iPad uses MIDI. This is good news because there are iOS applications that can monitor MIDI messages without needing an expensive hardware USB analayzer. I ordered the cheapest iPad I could find on eBay and will try reverse engineering the MIDI communication next week.

> work with RME products and many more like Focusrites

I don't think this is practical to combine in one program. AFAIK all the USB Audio Class compliant RME devices work similarly with a giant mixing matrix and could feasibly all be controlled with the same GUI application. But that is not the same as how the Focusrite Scarlett mixers work. Someone has already made a [dedicated program](https://github.com/geoffreybennett/alsa-scarlett-gui) for the Scarletts.

**Comment by agfline** [comment]
Yes, totalmix supports midi. If you're interested I can search my old HDDs for an arduino code I made to control totalmix (only the mixing part though).

Yes you're right, each constructor implements things its own way and most of the time very differently. However, they offer more or less the same functionalities (routing, mixing and a few fx). And with a well thought abstraction layer / functionalities description system, one could surely make a soft to rule them all. If you can build a totalmix FX for Linux that will be great step forward though !

Let me know when you have a repository ;)

**Comment by Be-ing** [comment]
> Let me know when you have a repository ;)

Will do! If you could contribute support for your device that would be great. Do you have the original FIreface UCX or the UCX II?

**Comment by agfline** [comment]
I have a [Fireface UC](https://www.rme-audio.de/fireface-uc.html), not UCX. There's no effect on the UC.
I don't have time right now to work diligently on a new project, but if you need anything I'll be happy to help.
